dagutils/traversal.go

```diff
@@ -0,0 +1,507 @@
+type Traversal struct {
+
+	// Each member of the slice is the parent of the following member, from
+	// the root to the current node, up until the position marked by `level`. The
+	// slice may contain more elements past that point but they should be
+	// ignored, every time the `level` increases the corresponding child node
+	// is inserted (the slice is not truncated to leverage the already allocated
+	// space). It should *always* have a length bigger than zero with the root
+	// of the DAG at the first position (empty DAGs are not valid).
+	path []ipld.Node
+
+	// This slice has the index of the child each parent in `path` is pointing
+	// to. Its current valid length is determined by `level`. The index in the
+	// parent can be set past all of its child nodes (having a value equal to
+	// `len(Links())`) to signal it has visited (or skipped) all of them. A
+	// leaf node with no children and its index in zero would also comply with
+	// this format.
+	childIndex []uint
+	// TODO: Revisit name, `childPosition`? Do not use the *link* term, put
+	// emphasis in the child (the *what*, not the *how* we get it).
+
+	// Depth of the node of the current position. It grows downwards, root is
+	// level 0, its child level 1, and so on. It controls the effective length
+	// of `path` and `childIndex`.
+	//
+	// A level of -1 signals the start case of a new `Traversal` that hasn't
+	// moved yet (or its position has been reset, see `ResetPosition`).
+	// Although this state is an invalid index to the slices, it allows to
+	// centralize all the visit calls in the `down` move (starting at zero
+	// would require a special visit case inside every traversal operation like
+	// `Iterate()` and `Search`). This value should never be returned to after
+	// the first `down` movement, moving up from the root should always return
+	// `ErrUpOnRoot`.
+	level int
+
+	// Method called each time a node is arrived upon in a traversal (through
+	// the `down()` movement, that is, when visiting it for the first time,
+	// not when going back up). It is the main API to implement functionality
+	// (like read and seek) on top of the `Traversal` structure.
+	//
+	// Its argument is the current `node` being visited. Any error it returns
+	// (apart from the internal `errDownStopIteration`) will be forwarded to
+	// the caller of the traversal operation (stopping it).
+	//
+	// Any of the exported methods of this API should be allowed to be called
+	// from within this method.
+	VisitHandler func(node ipld.Node) error
+
+	// Flag to stop the current iteration, intended to be set by the user
+	// inside `VisitHandler` to stop the current traversal operation.
+	Stop bool
+	// TODO: Use a function?
+
+	// Attribute needed to fetch nodes, will become really useful once node
+	// promises are implemented.
+	ctx  context.Context
+	serv ipld.NodeGetter
+
+	// The CID of each child of each node in the `path` (indexed by
+	// `level` and `childIndex`).
+	childCIDs [][]*cid.Cid
```

Comment:
that's gonna be a lot of memory, as is the promise set below

Reply:
This is also inherited from the `PBDagReader` (I'm glad this refactoring exposed some of the already existing characteristics of the current implementation) which has the

```Go
	// NodePromises for each of 'nodes' child links
	promises []*ipld.NodePromise

	// the cid of each child of the current node
	links []*cid.Cid
```

slices, each of those is instantiated in every level of the DAG when the recursive algorithm creates a new `PBDagReader`.

> lot of memory

Have in mind that `level` relates to the depth of the DAG (I should change its name as suggested above), so the size of this matrix grows at an `O(log n)` rate with the size of the DAG.

How would you advise this be handled?